src/goesproc/gradient.h

```diff
@@ -0,0 +1,170 @@
+      case 1: out.rgb[0] = q; out.rgb[1] = v; out.rgb[2] = p; break;
+      case 2: out.rgb[0] = p; out.rgb[1] = v; out.rgb[2] = t; break;
+      case 3: out.rgb[0] = p; out.rgb[1] = q; out.rgb[2] = v; break;
+      case 4: out.rgb[0] = t; out.rgb[1] = p; out.rgb[2] = v; break;
+      case 5: out.rgb[0] = v; out.rgb[1] = p; out.rgb[2] = q; break;
+    }
+
+    return out;
+  }
+};
+
+std::ostream& operator<<(std::ostream& os, const GradientPoint &p);
+
+enum GradientInterpolationType {
+  LERP_UNDEFINED = -1,
+  LERP_RGB,
+  LERP_HSV,
+  LERP_NUMTYPES
+};
+
+struct Gradient {
+  std::vector<GradientPoint> points;
+
+  inline int size() const {
+    return points.size();
+  }
+
+  // Search for the right place to insert a new point
+  int addPoint(const GradientPoint &point) {
+    auto i = points.begin();
+    while (i != points.end() && i->units < point.units) i++;
+    points.insert(i, point);
+    return size();
+  }
+
+  // Emit gradient points to stderr for debug purposes
+  void debug() const {
+    for (auto i = points.begin(); i != points.end(); i++) {
+      std::cerr << "[" << (i - points.begin()) << "] -> " << *i << std::endl;
+    }
+  }
+
+  // Find bounding points and perform linear interpolation
+  GradientPoint interpolate(const float units, const GradientInterpolationType lerptype = LERP_UNDEFINED) const {
+    auto left = points.end();
+    auto right = points.begin();
+
+    // Find left and right bounding points
+    auto i = points.begin();
+    for (i = i; i != points.end() && right == points.begin(); i++) {
+      if (i->units <= units) left = i;
+      if (i->units >= units) right = i;
+    }
+
+    // Clamp to ends of gradient if out of bounds
+    if (left == points.end()) return *(points.begin());
+    if (right == points.begin()) return *std::prev(points.end());
+
+    // Compute parametric distance between endpoints
+    auto du = right->units - left->units;
+    auto pd = (units - left->units) / du;
+
+    GradientPoint out;
+
+    if (lerptype == LERP_HSV) {
+      // Perform HSV interpolation
+      auto lh = left->hsv[0], rh = right->hsv[0];
+      double dh = rh - lh, adh = fabs(dh);
+
+      // Flip direction to choose shortest path in hue dimension
+      if (adh > 0.5) rh -= dh / adh;
+
+      float h = pd * rh + (1 - pd) * lh;
+      float s = pd * right->hsv[1] + (1 - pd) * left->hsv[1];
+      float v = pd * right->hsv[2] + (1 - pd) * left->hsv[2];
+
+      // Wrap the hue interpolation to deal with any negative values
+      if (h < 0.0) h += 1.0;
+
+      // Perform color-space conversion
+      out = out.fromHSV(h, s, v);
+    } else {
+      // Default to RGB interpolation unless user explicitly requests HSV
+      float r = pd * right->rgb[0] + (1 - pd) * left->rgb[0];
+      float g = pd * right->rgb[1] + (1 - pd) * left->rgb[1];
+      float b = pd * right->rgb[2] + (1 - pd) * left->rgb[2];
+
+      // Perform color-space conversion
+      out = out.fromRGB(r, g, b);
+    }
```

Comment:
... else throw error for invalid lerp type (downside of runtime configurability...).